### PR TITLE
[ADP-3320] Create skeleton HTTP API for Deposit Wallet

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request sketches how the `Cardano.Wallet.Deposit.HTTP` modules related to the `Cardano.Wallet.Deposit.IO` modules.

### Issue number

ADP-3320

### CODE REVIEW

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs
Comment (HeinrichApfelmus) [inline lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:1]: This module introduces the `ApiT` newtype that we want to use to model encodings at the HTTP API level — if `a` is a type, then a value of type `ApiT a` contains the same data as `a`, but it has an encoding (JSON, Bech32, …) that is specific to our HTTP API. We never expose a type `a` directly to the HTTP API, we only expose types of the form `ApiT a`.

In this module, the actual encoding (JSON, Bech32, …) is done in a style similar to the existing `cardano-wallet` codebase. However, in order to reduce boilerplate, we eventually want to use [FineTypes][1] to define the data types and get the encoding for free.

  [1]: https://github.com/cardano-foundation/fine-types

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Endpoints.hs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    ( liftIO
+    )
+import Servant.Server
+    ( Handler
+    )
+
+import qualified Cardano.Wallet.Deposit.IO as Wallet
+
+{-----------------------------------------------------------------------------
+    HTTP Wallet endpoints
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Endpoints.hs:33]: what about adding explicitly "principle" idea . So something like,
"
The exposed types go through an ApiT type that allows generating of JSON/CBOR instances for these wrapped types. It could be done by "by hand" or by reusing fine types instances 
"

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs
Comment (paweljakubas) [reply lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:1]: just put suggestion to document this "idea" somewhere above :+1:

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    Additional type definitions
+------------------------------------------------------------------------------}
+
+type CustomerList = [(Customer, Address)]
+
+{-----------------------------------------------------------------------------
+   ApiT
+------------------------------------------------------------------------------}
+
+newtype ApiT a = ApiT {unApiT :: a}
```
Comment (paweljakubas) [inline lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:66]: this might be located in the end somewhere else as it could not only refer to JSON ... maybe in `Cardano.Wallet.Deposit.HTTP.Types` module. `CustomerList` would go there too.

**Review by paweljakubas — APPROVED** [review]
very nice PR that shows the idea of how api/endpoint code is decomposed neatly. Very fond of openapi decomposition. Left some further suggestions. Also, testing of types, toy ones, against openapi spec to be added in separate PR? (could be in this or next PR of course). I bet we are going to test properties of `Wallet.listCustomer` and `Wallet.createAddress` and in integration testing just corner cases/regressions plus http errors or thin logic that is above `Wallet.*` functionality logic?

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:66]: Even though the name is general, I would like to keep `ApiT` specific to JSON and URL paths, so this module is a good place for now.

(I also want to avoid the situation where a child module `A.B.C` imports a parent module `A.B`.)

The `CustomerList` type is mainly a helpful abbreviation for use in the `*.HTTP.Types.API` module. If we really care about this definition, I would put it into `Cardano.Wallet.Deposit.IO` instead.

Comment (HeinrichApfelmus) [reply lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:66]: Oh, and I have to define the instances of `ApiT …` in the same module as the type `ApiT`, otherwise they are orphaned instances. 🤓

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Endpoints.hs (hunk 1 above)
Comment (HeinrichApfelmus) [reply lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Endpoints.hs:33]: Good idea. 👍  I have added a `NOTE [ApiT]`.

File: lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs
Comment (HeinrichApfelmus) [reply lib/customer-deposit-wallet/http/Cardano/Wallet/Deposit/HTTP/Types/JSON.hs:1]: done

**Comment by HeinrichApfelmus** [comment]
> Also, testing of types, toy ones, against openapi spec to be added in separate PR? (could be in this or next PR of course).

Yes, we want some automated testing that the JSON encoding satisfies the OpenAPI specification. However, with FineTypes, we can get that for free if we autogenerate the OpenAPI specification for the type definitions — we do not need to do any further testing if we assume that the FineTypes code is correct. This is the reason why I have not added any testing in this skeleton PR.

>  I bet we are going to test properties of `Wallet.listCustomer` and `Wallet.createAddress` and in integration testing just corner cases/regressions plus http errors or thin logic that is above Wallet.* functionality logic?

Exactly. 🤓 The main tests concerning functionality will be on `Cardano.Wallet.Deposit.IO` — including corner cases or regressions. The testing done at the `Cardano.Wallet.Deposit.HTTP` level will be restricted to JSON encodings, HTTP errors, or typos when plumbing `Cardano.Wallet.Deposit.IO` into `Cardano.Wallet.Deposit.HTTP` (e.g. accidentally omitted arguments or something). As far as the "official API" goes — that will be `Cardano.Wallet.Deposit.IO` and `Cardano.Wallet.Deposit.Pure`, depending on concern (with or without IO concerns).
